src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -0,0 +1,257 @@
+															   char *resultId);
+
+
+/*
+ * FindSubPlansUsedInPlan adds all the subplans used by the plan into
+ * plan->usedSubPlanNodeList field. Simply traversing the range table entries in the plan.
+ */
+void
+FindSubPlansUsedInPlan(DistributedPlan *plan)
+{
+	Query *jobQuery = NULL;
+	List *rangeTableList = NIL;
+	ListCell *rangeTableCell = NULL;
+
+	jobQuery = plan->workerJob == NULL ? plan->insertSelectSubquery :
+			   plan->workerJob->jobQuery;
+
+	/*
+	 * Decide if there are no intermediate results.
+	 *
+	 * We usually store the intermediate result info in workerJob field. We currently
+	 * have a single exception to this, which is SELECT INTO queries that have
+	 * ON CONFLICT/RETURNING clauses. If the query is of such a query we need to consider
+	 * pruning intermediate results even though we do not have a workerjob field
+	 *
+	 *  * We do not wish to proceed if one of the following is met:
+	 * - there are no intermediate results
+	 * - the query hits all the worker nodes
+	 */
+	if (jobQuery == NULL)
+	{
+		return;
+	}
+
+	rangeTableList = ExtractRangeTableEntryList(jobQuery);
+
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = lfirst(rangeTableCell);
+		if (rangeTableEntry->rtekind == RTE_FUNCTION)
+		{
+			Const *resultIdConst = NULL;
+			Datum resultIdDatum = 0;
+			char *resultId = NULL;
+
+			resultIdConst = FindResultIdOfIntermediateResultFunction(rangeTableEntry);
+
+			if (resultIdConst == NULL)
+			{
+				continue;
+			}
+
+			plan->usedSubPlanNodeList = list_append_unique(plan->usedSubPlanNodeList,
+														   resultIdConst);
+
+			resultIdDatum = resultIdConst->constvalue;
+			resultId = TextDatumGetCString(resultIdDatum);
+			elog(DEBUG4, "appending %s subplan to usedSubPlanNodeList of plan %lu",
+				 resultId, plan->planId);
+		}
+	}
+}
+
+
+/*
+ * RecordSubplanExecutionsOnNodes iterates over the usedSubPlanNodeList,
+ * and for each entry, record the workerNodes that are accessed by
+ * the distributed plan.
+ *
+ * Later, we'll use this information while we broadcast the intermediate
+ * results to the worker nodes. The idea is that the intermediate result
+ * should only be broadcasted to the worker nodes that are accessed by
+ * the distributedPlan(s) that the subPlan is used in.
+ *
+ * Finally, the function recursively descends into the actual subplans
+ * of the input distributedPlan as well.
+ */
+void
+RecordSubplanExecutionsOnNodes(HTAB *intermediateResultsHash,
+							   DistributedPlan *distributedPlan)
+{
+	List *usedSubPlanNodeList = distributedPlan->usedSubPlanNodeList;
+	ListCell *usedSubPlanNodeCell = NULL;
+	List *subPlanList = distributedPlan->subPlanList;
+	ListCell *subPlanCell = NULL;
+	int workerNodeCount = GetWorkerNodeCount();
+
+	foreach(usedSubPlanNodeCell, usedSubPlanNodeList)
+	{
+		Const *resultIdConst = lfirst(usedSubPlanNodeCell);
+		Datum resultIdDatum = resultIdConst->constvalue;
+		char *resultId = TextDatumGetCString(resultIdDatum);
```

Comment:
could we store a string in the list instead of Const? seems a bit odd to store the Const

Reply:
We can not use `list_append_unique()` for strings as the function checks for simple equality for pointers whereas deep equality is checked for numerous PostgreSQL structs such as `Const`

I am not so proud of this piece of code, but I found it hard to do it elegantly.

Reply:
`list_append_unique` does not work well with pointers. If I were to use strings here, I would also need several helper functions that check for uniqueness before appending, and I did not want to do that.

Reply:
Hmz, that's not ideal. Maybe we should try to change to a hash table.